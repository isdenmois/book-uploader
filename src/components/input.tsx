import React from 'react';
import {
  View,
  TextStyle,
  ViewStyle,
  StyleSheet,
  TextInput,
  Insets,
  TouchableOpacity,
  TouchableWithoutFeedback,
} from 'react-native';
import { useAutofocus } from 'utils/autofocus';
import * as colors from 'theme/colors';
import { ClearIcon } from './icons';

type Props = {
  value: string;
  onChange: (value: string) => void;
  icon: React.ReactNode;
  onSubmit?: () => void;
  disabled?: boolean;
  autoFocus?: boolean;
  initValue?: string;
  textColor?: string;
};

const hitSlop: Insets = { top: 10, right: 20, bottom: 10, left: 20 };

export function Input({ value, icon, onChange, onSubmit, disabled, initValue, autoFocus, textColor }: Props) {
  const inputRef = useAutofocus([initValue]);

  return (
    <TouchableWithoutFeedback testID='container' onPress={() => inputRef.current?.focus()}>
      <View style={s.container}>
        <View style={s.icon}>{icon}</View>

        <TextInput
          testID='input'
          style={[s.input, textColor && { color: textColor }]}
          editable={!disabled}
          value={value}
          onChangeText={onChange}
          onSubmitEditing={onSubmit}
          returnKeyType='search'
          placeholder='Search books'
          placeholderTextColor={colors.Search}
          ref={autoFocus && inputRef}
          autoFocus={autoFocus}
        />

        {!!value && (
          <TouchableOpacity testID='clear' style={s.clear} onPress={() => onChange('')} hitSlop={hitSlop}>
            <ClearIcon />
          </TouchableOpacity>
        )}
      </View>
    </TouchableWithoutFeedback>
  );
}

const s = StyleSheet.create({
  container: {
    flexDirection: 'row',
    margin: 15,
    marginHorizontal: 15,
    backgroundColor: colors.Input,
    borderRadius: 10,
    elevation: 4,
  } as ViewStyle,
  input: {
    flex: 1,
    fontSize: 14,
    paddingVertical: 6,
    paddingLeft: 10,
    color: colors.SearchText,
  } as TextStyle,
  icon: {
    marginLeft: 10,
    justifyContent: 'center',
  } as ViewStyle,
  clear: {
    paddingHorizontal: 8,
    paddingVertical: 5,
    justifyContent: 'center',
  } as ViewStyle,
});
