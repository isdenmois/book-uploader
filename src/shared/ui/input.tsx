import React, { MutableRefObject, useRef } from 'react'
import { TextInput, Insets, TouchableWithoutFeedback, TextInputProps, StyleSheet } from 'react-native'
import mergeRefs from 'react-merge-refs'
import { RecoilState, useRecoilState } from 'recoil'
import { useAutofocus } from 'shared/utils'

import { ClearIcon } from './icons'
import { Box, Theme, TouchableBox, useTheme } from './theme'

type Props = {
  state: RecoilState<string>
  icon: React.ReactNode
  onSubmit?: () => void
  disabled?: boolean
  initValue?: string
  textColor?: keyof Theme['colors']
  textInputRef?: MutableRefObject<TextInput>
} & TextInputProps

const hitSlop: Insets = { top: 10, right: 20, bottom: 10, left: 20 }

export function Input(props: Props) {
  const { state, icon, onSubmit, disabled, initValue, autoFocus, textColor, textInputRef, ...textInput } = props
  const inputRef = useAutofocus([initValue], autoFocus)
  const [value, onChange] = useRecoilState(state)
  const { colors } = useTheme()
  const clear = () => {
    onChange('')
    inputRef.current?.focus()
  }

  return (
    <TouchableWithoutFeedback testID='container' onPress={() => inputRef.current?.focus()}>
      <Box flexDirection='row' m={2} backgroundColor='input' alignItems='center' borderRadius={10} elevation={4}>
        <Box ml={1} justifyContent='center'>
          {icon}
        </Box>

        <TextInput
          testID='input'
          style={[s.input, textColor && { color: colors[textColor] }]}
          editable={!disabled}
          value={value}
          onChangeText={onChange}
          onSubmitEditing={onSubmit}
          returnKeyType='search'
          placeholderTextColor={colors.search}
          ref={mergeRefs([inputRef, textInputRef])}
          autoFocus={autoFocus}
          {...textInput}
        />

        {!!value && !disabled && (
          <TouchableBox paddingHorizontal={1} paddingVertical={0.5} testID='clear' onPress={clear} hitSlop={hitSlop}>
            <ClearIcon />
          </TouchableBox>
        )}
      </Box>
    </TouchableWithoutFeedback>
  )
}

export function useNextInput() {
  const next = useRef<TextInput>()
  const onSubmitEditing = () => next.current?.focus()

  return [next, onSubmitEditing] as const
}

const s = StyleSheet.create({
  input: {
    flex: 1,
    fontSize: 14,
    paddingVertical: 6,
    paddingLeft: 10,
  },
  clear: {
    paddingHorizontal: 8,
    paddingVertical: 5,
    justifyContent: 'center',
  },
})
