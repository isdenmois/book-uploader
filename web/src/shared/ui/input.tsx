import { Component, Show, JSXElement } from 'solid-js'
import { CloseIcon } from './icons'

interface Props {
  value: string
  onChange: (value: string) => void
  placeholder?: string
  disabled?: boolean
  type?: string
  icon?: JSXElement
  role?: any
}

export const Input: Component<Props> = props => {
  return (
    <div class='input flex relative items-center'>
      <span class='absolute inset-y-0 left-0 flex items-center pl-2' v-if='$slots.icon'>
        <Show when={!!props.icon}>{props.icon}</Show>
      </span>

      <input
        class='flex-1 py-2 rounded-md pl-10'
        placeholder={props.placeholder}
        value={props.value}
        type={props.type}
        disabled={props.disabled}
        onChange={(e: any) => props.onChange(e.target.value)}
        role={props.role}
      />

      <Show when={props.value && !props.disabled}>
        <div class='cursor-pointer px-2 z-10' onClick={() => props.onChange('')} data-testid='clear'>
          <CloseIcon />
        </div>
      </Show>
    </div>
  )
}
