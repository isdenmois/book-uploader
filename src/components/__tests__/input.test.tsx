import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { Text } from 'react-native';
import { Input } from '../input';
import { act } from 'react-test-renderer';

test('Input', () => {
  const onChange = jest.fn();
  const { toJSON, getByTestId, update } = render(<Input value='' onChange={onChange} icon={<Text />} />);

  expect(toJSON()).toMatchSnapshot();

  fireEvent(getByTestId('input'), 'onChangeText', 'test');

  expect(onChange).toHaveBeenCalledWith('test');

  act(() => update(<Input value='test' onChange={onChange} icon={<Text />} />));

  expect(toJSON()).toMatchSnapshot();

  fireEvent.press(getByTestId('clear'));
  fireEvent.press(getByTestId('container'));

  expect(onChange).toHaveBeenCalledWith('');
});
