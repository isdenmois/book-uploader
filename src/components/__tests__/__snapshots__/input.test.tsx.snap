// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 10,
      "elevation": 4,
      "flexDirection": "row",
      "margin": 15,
      "marginHorizontal": 15,
    }
  }
  testID="container"
>
  <View
    style={
      Object {
        "justifyContent": "center",
        "marginLeft": 10,
      }
    }
  >
    <Text />
  </View>
  <TextInput
    allowFontScaling={true}
    editable={true}
    onChangeText={[MockFunction]}
    placeholder="Search books"
    placeholderTextColor="#8C93A3"
    rejectResponderTermination={true}
    returnKeyType="search"
    style={
      Array [
        Object {
          "color": "#092D5D",
          "flex": 1,
          "fontSize": 14,
          "paddingLeft": 10,
          "paddingVertical": 6,
        },
        undefined,
      ]
    }
    testID="input"
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`Input 2`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 10,
      "elevation": 4,
      "flexDirection": "row",
      "margin": 15,
      "marginHorizontal": 15,
    }
  }
  testID="container"
>
  <View
    style={
      Object {
        "justifyContent": "center",
        "marginLeft": 10,
      }
    }
  >
    <Text />
  </View>
  <TextInput
    allowFontScaling={true}
    editable={true}
    onChangeText={
      [MockFunction] {
        "calls": Array [
          Array [
            "test",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    placeholder="Search books"
    placeholderTextColor="#8C93A3"
    rejectResponderTermination={true}
    returnKeyType="search"
    style={
      Array [
        Object {
          "color": "#092D5D",
          "flex": 1,
          "fontSize": 14,
          "paddingLeft": 10,
          "paddingVertical": 6,
        },
        undefined,
      ]
    }
    testID="input"
    underlineColorAndroid="transparent"
    value="test"
  />
  <View
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 10,
        "left": 20,
        "right": 20,
        "top": 10,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 5,
      }
    }
    testID="clear"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={24}
      bbWidth={24}
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      vbHeight={512}
      vbWidth={320}
      width={24}
    >
      <RNSVGGroup>
        <RNSVGPath
          d="M207.6 256l107.72-107.72c6.23-6.23 6.23-16.34 0-22.58l-25.03-25.03c-6.23-6.23-16.34-6.23-22.58 0L160 208.4 52.28 100.68c-6.23-6.23-16.34-6.23-22.58 0L4.68 125.7c-6.23 6.23-6.23 16.34 0 22.58L112.4 256 4.68 363.72c-6.23 6.23-6.23 16.34 0 22.58l25.03 25.03c6.23 6.23 16.34 6.23 22.58 0L160 303.6l107.72 107.72c6.23 6.23 16.34 6.23 22.58 0l25.03-25.03c6.23-6.23 6.23-16.34 0-22.58L207.6 256z"
          fill={4288391370}
          propList={
            Array [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
